feat: AI-powered summary generation for notes

ANINDITAPRIYADARSHINI opened the pull request:
## What
Implemented AI-powered TL;DR summary generation for notes using DigitalOcean's Gradient AI, allowing users to quickly absorb note content.

## Why
Users need a quick way to understand the key points of their notes without reading the full content. AI-generated summaries provide instant TL;DR versions that help with quick note review, content scanning, and better note organization.

## How

### Backend Implementation
- Added `generateSummary()` method to `DigitalOceanInferenceService` class
- Added `summary` field to `Note` model (Prisma schema)
- Created `/api/notes/[id]/generate-summary` route (POST: generate, DELETE: clear)
- Created `/api/user/preferences` endpoint for global settings

### Frontend Implementation
- Added "Generate Summary" button in note edit/view modes
- Summary displayed prominently at top with "TL;DR Summary" label
- "Regenerate" button to generate new summary
- "Clear" button to remove unwanted summaries
- Global AI Summary Settings toggle in account page
- Toast notifications for user feedback


## User Flow
1. User opens existing note in edit/view mode
2. Click "Generate Summary" button below content
3. AI generates TL;DR summary displayed at top of note
4. User can click "Regenerate" for a new summary
5. User can click "Clear" to remove summary
6. Settings page allows global enable/disable via toggle



## Acceptance Criteria

✅ Button to trigger summary generation  
✅ UI shows summary prominently at top of note  
✅ Users can refresh/regenerate summary  
✅ Disable per note (clear button)  
✅ Disable globally (settings toggle)  

## Screenshots

### Core Functionality

**Generate Summary Button**
<img width="1364" height="640" alt="generate_summary_button" src="https://github.com/user-attachments/assets/421340d3-ba83-4994-9e14-981dc582e090" />


**Summary Generation - Loading State**
<img width="717" height="421" alt="generating_summary" src="https://github.com/user-attachments/assets/8afbc145-42a8-4ef5-914c-f9906bc366e7" />


**Generated Summary at Top with Regenerate & Clear**
<img width="720" height="577" alt="Generated summary" src="https://github.com/user-attachments/assets/710da29f-e97e-4fc9-8ef9-6ae4aabb5b70" />


### Summary Actions

**Regenerate Functionality**
<img width="1366" height="642" alt="regenerated summary" src="https://github.com/user-attachments/assets/1e3677e1-146b-48ef-abfd-a9a648fc6e2f" />


**Clear Functionality**
<img width="1366" height="639" alt="Cleared summary" src="https://github.com/user-attachments/assets/ecb57175-f8d9-4ee6-a8f2-6550266f322a" />


### Settings Integration

**Settings - AI Summaries Enabled**
<img width="1365" height="640" alt="AI summary settings globally" src="https://github.com/user-attachments/assets/d6fe0957-c420-4cf8-87f6-53b98e401751" />


**Settings - AI Summaries Disabled**
<img width="881" height="171" alt="toogle feature off" src="https://github.com/user-attachments/assets/c9cbfa90-54de-41bf-8bf0-c5bb6fb9ff8b" />


### Full Context View

**Complete Note View with Summary**
<img width="720" height="580" alt="Note_with_summary" src="https://github.com/user-attachments/assets/43b0faa5-39a7-4773-bda5-7da61a7733e2" />


Fixes #30

ANINDITAPRIYADARSHINI committed:
feat: add generateSummary method to AI service

- Added generateSummary() method to DigitalOceanInferenceService
- Generates 1-3 sentence TL;DR summaries from note content
- Uses Gradient AI with appropriate system prompt
- Follows existing service patterns for title/content generation

ANINDITAPRIYADARSHINI committed:
feat: add summary field to Note schema

- Added optional summary field to store AI-generated summaries
- Database migration will be needed (maintainers will run)

ANINDITAPRIYADARSHINI committed:
feat: add API endpoint for summary generation

- Created POST /api/notes/[id]/generate-summary endpoint
- Handles authentication and note ownership verification
- Calls AI service to generate summary
- Updates note with generated summary in database

ANINDITAPRIYADARSHINI committed:
feat: add summary generation UI to note form

- Added Generate Summary button in edit/view modes
- Display summary prominently at top with TL;DR label
- Integrated with AI summary generation API
- Added loading states and toast notifications
- Follows existing Material UI design patterns

ANINDITAPRIYADARSHINI committed:
feat: add clear summary functionality

- Added Clear button in summary display
- Implemented DELETE endpoint to remove summaries
- Users can now remove unwanted summaries from notes

ANINDITAPRIYADARSHINI committed:
feat: add global AI summary toggle in settings

- Created AISummarySettings component with toggle switch
- Added to account settings page
- Implemented preferences API endpoint (GET/PUT)
- Users can now disable AI summary features globally
- Includes helpful description and disabled state alert

ANINDITAPRIYADARSHINI committed:
fix: update imports to work with Next.js 15 and local auth setup

- Fixed auth imports to use relative paths
- Updated to use auth() instead of getServerSession
- Fixed params access to await in Next.js 15
- Applied database migration for summary field

Haimantika commented:
> Hello @Haimantika,
> 
> I am following up on the final review status. My feature implementation is complete and ready.
> 
> To demonstrate the AI features, I need a DigitalOcean Inference API key. However, DigitalOcean requires a payment method to generate the API key, which I don't have. Could you please provide a **test key** so I can complete the required demonstration and move this PR forward?
> 
> I also want to bring to your attention that a duplicate Pull Request (PR #48) has been opened. **Resolving the API key request now will allow us to merge my original PR quickly.**
> 
> Thank you for your assistance.

Hi @ANINDITAPRIYADARSHINI you should be able to get one by creating an account on DigitalOcean and use the $200 credits for it. Please try that first. If you are not able to then send me your email and I will share a temporary key.

Haimantika commented:
> Hi @Haimantika! I tried signing up but didn't receive any $200 credits.
> 
> **Here is my email:** [anindita.priyadarshini48@gmail.com](mailto:anindita.priyadarshini48@gmail.com)
> 
> Please share the temporary API key so I can demonstrate the feature.
> 
> Thanks!

What problem are you facing?

ANINDITAPRIYADARSHINI committed:
feat: add regenerate button, move summary to top, fix delete error

- Added Regenerate button next to Clear in summary box
- Moved summary display to top of note (below title, above content)
- Fixed delete note error handling to properly parse response
- Improved spacing between summary and content fields

ANINDITAPRIYADARSHINI committed:
feat: complete AI summary feature implementation

- Added Regenerate button alongside Clear button in summary display
- Moved summary display to top of note
- Fixed delete note error handling
- Added success/error toast notifications
- Implemented settings toggle UI for global enable/disable
- Improved spacing between UI elements

Note: Settings persistence requires backend implementation

ANINDITAPRIYADARSHINI commented:
> > Hello @Haimantika,
> > 
> > Implementation completed! All acceptance criteria met and tested.
> > 
> > Screenshots attached demonstrating the full workflow. Ready for review.
> > 
> > Thanks!
> 
> Hi @ANINDITAPRIYADARSHINI great work! I have added the hacktoberfest-accepted label that should help. We will merge the PR after we have done more checks.

Thank you @Haimantika ! Looking forward to it.

ANINDITAPRIYADARSHINI committed:
Merge branch 'main' into feat/ai-summary-generation

ANINDITAPRIYADARSHINI committed:
Merge branch 'main' into feat/ai-summary-generation

Haimantika commented:
> Hello @Haimantika,
> 
> Following up on this PR. It's been reviewed and already accepted by Hacktoberfest, but still open after so many days.
> 
> May I know what's the timeline for merging, and by when can I expect the PR to be closed?
> 
> Would appreciate an update.
> 
> Thanks!

Hi @ANINDITAPRIYADARSHINI please give us some more time for this. We will try to get it merged by February, and if you have any questions you can email me directly.

ANINDITAPRIYADARSHINI commented:
> Hi @ANINDITAPRIYADARSHINI please give us some more time for this. We will try to get it merged by February, and if you have any questions you can email me directly.

Hi @Haimantika Thank you for the update. I'll wait for the merge by February!

Haimantika commented on application/prisma/migrations/20251011134106_add_summary_field/migration.sql:
```diff
@@ -0,0 +1,11 @@
+ALTER TABLE "Note" ADD COLUMN     "summary" TEXT;
+
+-- CreateIndex
+CREATE UNIQUE INDEX "Subscription_userId_key" ON "Subscription"("userId");
```
@ANINDITAPRIYADARSHINI I was testing it locally. This is the main part.

It should look something like: ALTER TABLE "Note" ADD COLUMN IF NOT EXISTS "summary" TEXT;

-- CreateIndex
CREATE UNIQUE INDEX IF NOT EXISTS "Subscription_userId_key" ON "Subscription"("userId");

Haimantika commented on application/src/services/ai/digitalOceanInferenceService.ts:
This model is deprecated. Use something like: llama3-8b-instruct

Haimantika commented:
@ANINDITAPRIYADARSHINI once you work on the above two comments from my end. We can merge it.

Great job on the PR and thanks for your patience. I tested it and it works as expected.
<img width="1032" height="782" alt="Screenshot 2026-03-02 at 1 28 47 PM" src="https://github.com/user-attachments/assets/1aeca0dd-9078-4ed7-9dbe-37728fcedae4" />

ANINDITAPRIYADARSHINI committed:
Add summary column and unique index if not exists

Fix migration to use IF NOT EXISTS checks

ANINDITAPRIYADARSHINI committed:
Change model from 'anthropic-claude-3-opus' to 'llama3-8b-instruct'

ANINDITAPRIYADARSHINI commented:
@Haimantika Thank you! I've addressed both comments and pushed the changes. The PR is ready for merge now. I really appreciate you testing it and thanks once again for all your guidance throughout this contribution!

> @ANINDITAPRIYADARSHINI once you work on the above two comments from my end. We can merge it.
> 
> Great job on the PR and thanks for your patience. I tested it and it works as expected. <img alt="Screenshot 2026-03-02 at 1 28 47 PM" width="1032" height="782" src="https://private-user-images.githubusercontent.com/32809211/556769905-1aeca0dd-9078-4ed7-9dbe-37728fcedae4.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzI0NDIwNzUsIm5iZiI6MTc3MjQ0MTc3NSwicGF0aCI6Ii8zMjgwOTIxMS81NTY3Njk5MDUtMWFlY2EwZGQtOTA3OC00ZWQ3LTlkYmUtMzc3MjhmY2VkYWU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAzMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMzAyVDA4NTYxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExYWEwMTU3YzM0NDBlMTEzMGUzMmQwMzI1MmQwNTc1OWNlZjM3MjU2YTk0NjQzZDQ1YWEyZTUxNzBhMjU3NGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xEL2gFhcBCo6MJA0vRpAg7GakGC3U_7bSJf4lKtWaDE">

@Haimantika Thank you! I've addressed both comments and pushed the changes. The PR is ready for merge now. I really appreciate you testing it and thanks once again for all your guidance throughout this contribution!

Haimantika commented:
> @Haimantika Thank you! I've addressed both comments and pushed the changes. The PR is ready for merge now. I really appreciate you testing it and thanks once again for all your guidance throughout this contribution!
> 
> > @ANINDITAPRIYADARSHINI once you work on the above two comments from my end. We can merge it.
> > Great job on the PR and thanks for your patience. I tested it and it works as expected. <img alt="Screenshot 2026-03-02 at 1 28 47 PM" width="1032" height="782" src="https://private-user-images.githubusercontent.com/32809211/556769905-1aeca0dd-9078-4ed7-9dbe-37728fcedae4.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzI0NDIwNzUsIm5iZiI6MTc3MjQ0MTc3NSwicGF0aCI6Ii8zMjgwOTIxMS81NTY3Njk5MDUtMWFlY2EwZGQtOTA3OC00ZWQ3LTlkYmUtMzc3MjhmY2VkYWU0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAzMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMzAyVDA4NTYxNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExYWEwMTU3YzM0NDBlMTEzMGUzMmQwMzI1MmQwNTc1OWNlZjM3MjU2YTk0NjQzZDQ1YWEyZTUxNzBhMjU3NGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.xEL2gFhcBCo6MJA0vRpAg7GakGC3U_7bSJf4lKtWaDE">
> 
> @Haimantika Thank you! I've addressed both comments and pushed the changes. The PR is ready for merge now. I really appreciate you testing it and thanks once again for all your guidance throughout this contribution!

Looks good to me!
